### PR TITLE
Access spawner change: null default access on doors

### PR DESCRIPTION
**Pull request by Kubius**
[QOL][REWORK]
## About the PR 

Modifies access spawners and airlocks to make access spawners behave more "predictably". Airlocks are given a "mod_by_spawner" flag that starts as 0 - when an access spawner procs, it will now check for this flag if it's modifying an airlock, and null the existing access before adding its own access and setting the flag to 1. This permits airlocks to retain default access in their base type while also making it so said default access doesn't accidentally "ride along" unless explicitly disabled with var-editing.

I've tagged this as a rework because it alters the existing functionality, and may disrupt maps if they've been configured relying on the old behavior.

## Why's this needed? 

Access spawners not nulling existing access continually causes issues with incorrect permissions. After this change, the access a mapper sees in the spawners will be precisely what they get.

### CODE REVIEW

File: code/obj/access_spawn.dm [hunk 1]
```diff
@@ -22,6 +22,9 @@
 
 	proc/setup()
 		for (var/obj/machinery/M in src.loc)
+			if (istype(M,/obj/machinery/door/airlock) && M:modBySpawner == 0) //access spawn mapper QoL upgrade: no more accidental extra access
+				M.req_access = null
+				M:modBySpawner = 1
```
Comment (Adharainspace) [inline code/obj/access_spawn.dm:27]:
```suggestion
			if (istype(M, /obj/machinery/door/airlock)) //access spawn mapper QoL upgrade: no more accidental extra access
				var/obj/machinery/door/airlock/A = M
				if (!A.modded_by_spawner)
					A.req_access = null
					A.modded_by_spawner = 1
```
changes it up to not use the : as its generally bad practice. instead you should typecheck, set a new var thats of the type you want, and then use that one.
also changes the var name to be more in line with our existing conventions

File: code/obj/machinery/door/airlock.dm [hunk 2]
```diff
@@ -102,6 +102,9 @@ Airlock index -> wire color are { 9, 4, 6, 7, 5, 8, 1, 2, 3 }.
 
 	var/no_access = 0
 
+	//for access spawn mapper QoL upgrade
+	var/modBySpawner = 0
```
Comment (Adharainspace) [inline code/obj/machinery/door/airlock.dm:106]:
```suggestion
	var/modded_by_spawner = 0 //has this airlock had its access changed by a spawner? 0 if no, 1 if yes
```
edits var name to be more in line with naming conventions, adds a comment explaining it

**Comment by Adharainspace** [comment]
although i have to wonder if this var is even necessary. why not just null access on all airlocks when an access spawner is placed on them? or do you have to stack spawners?

**Comment by Kubius** [comment]
The var is necessary, as a combination of access spawners (i.e. chef and barman) can be used, and it lets you do that.

As to the use of : in the evaluation - I'm of the opinion that while it's a risky tool to use, it's a valid use case because the check for it working is immediately adjacent and very readable, and it saves having to allocate another variable and move the reference over to it (only a tiny performance gain, but it's seemingly at no cost and given it's done 200+ times a map it might multiply out enough to matter a bit). I'll swap it out if there's a more broad inclination that it should be swapped, or if I'm misinterpreting the performance difference.

**Comment by mordent-goonstation** [comment]
Access spawners are significantly more visible than some random var on the machine underneath. If the goal is to make it so that you can combine them in interesting ways, I would look to improve how the spawners interact if both placed rather than inverting control like this.

I think the "problem" that is trying to be solved is simply going to be moved to a different one instead if this is merged (i.e. people forgetting to set the flag). It's hard to tell at a glance which doors do/don't have that flag set, whereas it's easy to see where the access spawners are, so I think this is a step in the wrong direction.

**Comment by mordent-goonstation** [comment]
> As to the use of : in the evaluation - I'm of the opinion that while it's a risky tool to use, it's a valid use case because the check for it working is immediately adjacent and very readable, and it saves having to allocate another variable and move the reference over to it (only a tiny performance gain, but it's seemingly at no cost and given it's done 200+ times a map it might multiply out enough to matter a bit). I'll swap it out if there's a more broad inclination that it should be swapped, or if I'm misinterpreting the performance difference.

Firmly disagree. I'd rather have the map build for another second or two than have to worry about `:` usage, even assuming it adds that amount of overhead. It's "clear" for now, but nothing is stopping people from messing with the surrounding code, blissfully unaware of issues it can cause.

**Comment by Adharainspace** [comment]
ok so a suggestion mordent had was to just delete all existing instances of doors having preset access vars, and then nixing the rest of this pr. avoids the issue, 1 less var on airlocks

**Comment by Kubius** [comment]
"I think the "problem" that is trying to be solved is simply going to be moved to a different one instead if this is merged (i.e. people forgetting to set the flag)."

The access spawner sets the flag itself - not sure quite what you mean here.

**Comment by mordent-goonstation** [comment]
I was assuming the case where you want the underlying door type's access + whatever access you put on top, rather than what this seems to do (which is make it so that if you have an access spawner on the door the underlying door's access levels get nulled, then all of the spawners apply their access requirements).

Have you done any checks on what doors are going to need fixing because of this change?

**Comment by Kubius** [comment]
Did a brief assessment of impact on rotation maps:

- Atlas: Mix of access spawners and var-edits, but all manually entered access is via req_access_txt. Should be safe?
- Cog1: 1 access spawner on a crate; crates aren't affected. Safe.
- Cog2: No access spawners. Safe.
- Destiny: No access spawners. Safe.
- Donut3: Full access spawners, including maintenance doors. Should be safe.
- Horizon: Mix of access spawners and var-edits. Focused primarily on areas at the edge of maintenance and didn't spot any instances of excess/missing access that would end up problematic, though found some instances of incorrect access levels applied (not a problem caused by this). Probably safe.
- Kondaru: Access spawners, non-spawner doors left default - the only req_access edits are manual nulling. Should be safe.
- Manta: Full access spawners, including maintenance doors. Should be safe.
- Oshan: Access spawners on most doors, and doors without access spawners appear to be using their default access "as designed". Should be safe.

**Comment by Kubius** [comment]
Pali will evidently be doing the other approach (remove default access, add spawners) so I'm closing this for the time being.

**Pull request closed without merging**